include/fluidsynth/synth.h

```diff
@@ -46,6 +46,97 @@ extern "C" {
  */
 
 
+/*******************************************************************
+ API: Poly mono mode 
+ */
+
+/* Macros interface to poly/mono mode variables */
+enum fluid_basic_channel_mode_flags
+{
+    MONO = 0x01, /* bit 0, 0: poly on , 1: mono on */
+    OMNI = 0x02, /* bit 1, 0: omni on, 1:omni off */
+    BASIC_CHANNEL = 0x04, /* bit 2, 1: channel is basic channel */
+    ENABLED = 0x08,         /* bit 3, 1: channel is listened */
```

Comment:
I think it's a bit misleading that MONO and OMNI are switched on if the bit is 0, BASIC_CHANNEL and ENABLED are switched on if the bit is 1. I would vote to make them all behave in the same way: enabled if bit is set, otherwise disabled.

Also a small typo here "Channel is listened" doesn't really make sense. Maybe "channel is enabled"?

Reply:
> I think it's a bit misleading that MONO and OMNI are switched on if the bit is 0, 

Sorry, that only applies to OMNI, of course. MONO is already on if the bit is set.

Reply:
If I understood correctly, this was made because by default fluidsynth has one basic channel
on MIDI channel 0 with PolyOn and OmniOn, so that the synth plays/responds to all MIDI data received on all 16 channels.

Reply:
Though I agree it would be clearer to have this bit flipped.

Reply:
MIDI spec `channel mode (1 to 4)` are in Fluidsynth zero based (0 to 3) respectively.
This value is coded internally on 2 bits (bits b1,b0) and simply represent the four MIDI spec values (0 based):
- OMNION_POLY  : mode 0
- OMNION_MONO: mode 1
- OMNIOFF_POLY: mode 2
- OMNIOFF_MONO: mode 3 

Consequently, MONO On is represented by bit 0 when 1, and OMNI ON is represented by bit b2 when 0. 

>by default fluidsynth has one basic channel on MIDI channel 0 with PolyOn and OmniOn, so that the synth plays/responds to all MIDI data received on all 16 channels.

Yes, mode 0 on channel 0 is the preferable default mode(and usually recommended by MIDI spec).

>Also a small typo here "Channel is listened" doesn't really make sense. Maybe "channel is enabled"?

Yes, it could be done

Reply:
> Consequently, MONO On is represented by bit 0 when 1, and OMNI ON is represented by bit b2 when 0.

Ok, if it is important to have the same numbering of modes inside of FluidSynth as used in the spec, then maybe we could rename the enums to better reflect what they express. Maybe POLY_OFF and OMNI_OFF.

The current state is really quite confusing, with OMNI being the only one where the logic is reversed. What tripped me up was the following code:

```
if (chan->mode &  BASIC_CHANNEL)
{
    int new_mode, new_val;
--> if(!(chan->mode & OMNI)) /* channel is actually Omni On */
    {new_mode = OMNION_MONO; new_val=0;} /* set channel in mode 1 */
    else /* channel is actually Omni Off */
    {new_mode = OMNIOFF_MONO; new_val= value;} /* set channel in mode 3 */
    return fluid_synth_set_basic_channel_LOCAL(synth,channum, 
                                        new_mode, new_val);
}
```

So we have ```(mode & BASIC_CHANNEL)``` to mean "is basic channel" and ```(mode & OMNI)``` to mean "is NOT omni". This is really confusing and a great source for errors. If we would rename MONO and OMNI to POLY_OFF and OMNI_OFF, then those bit tests suddenly become meaningful and consistent again.

Reply:
>Ok, if it is important to have the same numbering of modes inside of FluidSynth as used in the spec, then maybe we could rename the enums to better reflect what they express. Maybe POLY_OFF and OMNI_OFF.


Yes, the naming should be changed `POLY_OFF and OMNI_OFF` or `MONO_ON and OMNI_OFF` ?  which did  you prefer ?.

Reply:
`POLY_OFF and OMNI_OFF`

Reply:
I personally would prefer POLY_OFF and OMNI_OFF, makes it more consistent.

Reply:
So, `MONO , OMNI` will be changed  to `POLY_OFF`, `OMNI_OFF` respectively.

Reply:
Then I would suggest also renaming the other member of that enum to add ```_BIT```, e.g. BASIC_CHANNEL_BIT, ENABLED_BIT

My preference would be to prefix everything with CHANNEL_:
```
CHANNEL_POLY_OFF
CHANNEL_OMNI_OFF
CHANNEL_BASIC
CHANNEL_ENABLED
```

Or ```CHAN_POLY_OFF```, if you think CHANNEL_ is too long.

Reply:
> Or CHAN_POLY_OFF, if you think CHANNEL_ is too long.

Because it's in a public header, we should also prefix `FLUID_`. I would also like to put the `BREATH_*` flags into that enum:

```
enum fluid_basic_channel_mode_flags
{
    FLUID_CHANNEL_POLY_OFF = 0x01, /* bit 0, 0: poly on , 1: mono on */
    FLUID_CHANNEL_OMNI_OFF = 0x02, /* bit 1, 0: omni on, 1:omni off */
    FLUID_CHANNEL_BASIC = 0x04, /* bit 2, 1: channel is basic channel */
    FLUID_CHANNEL_ENABLED = 0x08,         /* bit 3, 1: channel is listened */

    FLUID_CHANNEL_BREATH_POLY = 0x10,
    FLUID_CHANNEL_BREATH_MONO = 0x20,
    FLUID_CHANNEL_BREATH_SYNC = 0x40,

    FLUID_CHANNEL_MODE_MASK,
    FLUID_CHANNEL_MODE_OMNION_POLY,
    FLUID_CHANNEL_MODE_OMNION_MONO,
    ...
    FLUID_CHANNEL_MODE_LAST // NOT PART OF PUBLIC API/ABI stability guarantee
};
```

If you agree I would take care of these changes.

Reply:
>Because it's in a public header, we should also prefix FLUID_. I would also like to put the BREATH_* flags into that enum:

Yes, these  missing flags need to be added.

>and also fix logic of POLY_OFF bit then.

Keeping the same numbering of `mode value` inside of FluidSynth as used in the `spec` simplify the setting/getting of this `mode value` in API `fluid_synth_set_basic_channel(), fluid_synth_get_basic_channels(), and fluid_synth_get_channel_mode()`. 
Changing the logic of bits `(b1,b0)` will complicate the construction of  `MIDI spec. mode` from the `FluidSynth mode b1,b0`  inside `fluid_synth_get_basic_channels(), and fluid_synth_get_channel_mode()`. This also will complicate the construction of `FluidSynth mode b1,b1` from the `MIDI spec. mode` inside `fluid_synth_set_basic_channel()`. So what could be chosen ?

Reply:
>> and also fix logic of POLY_OFF bit then.

Sorry, I was confused. The logic will be fine ofc, once the flags have been renamed. So if you agree with the re-naming I'd take care of this then.

Reply:
>So if you agree with the re-naming I'd take care of this then.
Yes ofc, these missing flags BREATH_* need to be added and also the re-naming is necessary.